core/node/node_framework/src/service/context.rs

```diff
@@ -98,4 +90,15 @@ impl<'a> ServiceContext<'a> {
     pub async fn get_resource_or_default<T: Resource + Clone + Default>(&mut self) -> T {
         self.get_resource_or_insert_with(T::default).await
     }
+
+    /// Adds a resource to the service.
+    /// If the resource with the same name is already provided, the method will return an error.
+    pub fn add_resource<T: Resource>(&mut self, resource: T) -> Result<(), WiringError> {
```

Comment:
Ah, inserting resources makes the `ServiceContext` container look more like a traditional DI container (IMO, that's a good thing). 

- Bikeshedding: I feel that per Rust naming conventions, `insert_resource` would be more idiomatic (cf. map / set methods in `std`).
- Here and in other `ServiceContext` methods, it would make sense to log operations to make potential DI problems easier to investigate. Could be done later, of course.

Reply:
> Could be done later, of course.

Yup, I'm currently thinking about the right logging structure that would keep track of the initialization process (e.g. "Wiring Layer X requested resource Y"), but I want to tackle it later.